packages/medusa/src/loaders/module.ts

```diff
@@ -0,0 +1,56 @@
+import { asFunction, asValue } from "awilix"
+import { trackInstallation } from "medusa-telemetry"
+import { ConfigModule, Logger, MedusaContainer } from "../types/global"
+import { ModulesHelper } from "../utils/module-helper"
+
+type Options = {
+  container: MedusaContainer
+  configModule: ConfigModule
+  logger: Logger
+}
+
+const ModuleHelper = new ModulesHelper()
+
+export default async ({
+  container,
+  configModule,
+  logger,
+}: Options): Promise<void> => {
+  const moduleResolutions = configModule?.moduleResolutions ?? {}
+
+  for (const resolution of Object.values(moduleResolutions)) {
+    try {
+      const loadedModule = await import(resolution.resolutionPath!)
+
+      const moduleLoaders = loadedModule?.loaders || []
+      for (const loader of moduleLoaders) {
+        await loader({ container, configModule, logger })
+      }
+
+      const moduleServices = loadedModule?.services || []
+
+      for (const service of moduleServices) {
+        container.register({
+          [resolution.definition.registrationName]: asFunction(
+            (cradle) => new service(cradle, configModule)
+          ).singleton(),
+        })
+      }
+
+      const installation = {
+        module: resolution.definition.key,
+        resolution: resolution.resolutionPath,
+      }
+
+      trackInstallation(installation, "module")
+    } catch (err) {
+      console.log("Couldn't resolve module: ", resolution.definition.label)
```

Comment:
thought: If the module is required we should crash the bootstrap here

Reply:
See [18a70dc](https://github.com/medusajs/medusa/pull/2597/commits/18a70dcd48a74ccbe617e8391b48adcbc5b83fef)
Merge state
Add more commits by pushing to the [feat/module-resolutions](https://github.com/medusajs/medusa/tree/feat/module-resolutions) branch on [medusajs/medusa](https://github.com/medusajs/medusa).

Code owner review required
Waiting on code owner review from medusajs/core and/or medusajs/docs. 
1 approval
Some checks haven’t completed yet
1 queued, 5 in progress, 3 expected, 1 neutral, and 1 successful checks
@github-actions
Medusa Pipeline / setup (pull_request) Queued — Waiting to run this check...
[Details](https://github.com/medusajs/medusa/actions/runs/3487177254/jobs/5834494632)
@github-actions
CLI Pipeline / test-cli-with-database (sqlite) (pull_request) In progress — This check has started...
Required
[Details](https://github.com/medusajs/medusa/actions/runs/3487177253/jobs/5834494603)
@github-actions
Documentation Tests / docs-test (pull_request) In progress — This check has started...
[Details](https://github.com/medusajs/medusa/actions/runs/3487177250/jobs/5834494544)
@github-actions
OAS Comments Format Validation / docs-test (pull_request) In progress — This check has started...
[Details](https://github.com/medusajs/medusa/actions/runs/3487177256/jobs/5834494653)
@github-actions
CLI Pipeline / test-cli-with-database (postgres) (pull_request) In progress — This check has started...
Required
[Details](https://github.com/medusajs/medusa/actions/runs/3487177253/jobs/5834494725)
@github-actions
Documentation Tests / lint (pull_request) In progress — This check has started...
[Details](https://github.com/medusajs/medusa/actions/runs/3487177250/jobs/5834494778)
integration-tests-api Expected — Waiting for status to be reported
Required
integration-tests-plugins Expected — Waiting for status to be reported
Required
unit-tests Expected — Waiting for status to be reported
Required
@kodiakhq
kodiakhq: status — Ignored (no automerge label: 'automerge')
[Details](https://github.com/medusajs/medusa/pull/2597/checks?check_run_id=9546753203)
@gitguardian
GitGuardian Security Checks Successful in 2s — No secrets detected ✅
[Details](https://github.com/medusajs/medusa/pull/2597/checks?check_run_id=9546750643)
This branch is out-of-date with the base branch
Merge the latest changes from develop into this branch.
 Merge without waiting for requirements to be met (bypass branch protections)

 You can also [open this in GitHub Desktop](https://desktop.github.com/) or view .
@olivermrbl

 
Add heading textAdd bold text, <Cmd+b>Add italic text, <Cmd+i>
Add a quote, <Cmd+Shift+.>Add code, <Cmd+e>Add a link, <Cmd+k>
Add a bulleted list, <Cmd+Shift+8>Add a numbered list, <Cmd+Shift+7>Add a task list, <Cmd+Shift+l>
Directly mention a user or team
Reference an issue, pull request, or discussion
Add saved reply
Leave a comment
No file chosen
Attach files by dragging & dropping, selecting or pasting them.
Styling with Markdown is supported
Remember, contributions to this repository should follow its [contributing guidelines](https://github.com/medusajs/medusa/blob/3aa082714bc9a1df47e5a5db9e542084d2dd63bb/CONTRIBUTING.md) and [security policy](https://github.com/medusajs/medusa/blob/3aa082714bc9a1df47e5a5db9e542084d2dd63bb/SECURITY.md).
 ProTip! Add [.patch](https://github.com/medusajs/medusa/pull/2597.patch) or [.diff](https://github.com/medusajs/medusa/pull/2597.diff) to the end of URLs for Git’s plaintext views.
Reviewers
@[srindom](https://github.com/srindom)
srindom
@[adrien2p](https://github.com/adrien2p)
adrien2p
At least 1 approving review is required to merge this pull request.

Still in progress? 
Assignees
No one—
Labels
None yet
Projects
None yet
Milestone
No milestone
Development
Successfully merging this pull request may close these issues.

None yet

Notifications
Customize
You’re receiving notifications because you’re watching this repository.
5 participants
@olivermrbl
@srindom
@adrien2p
@fPolic
@carlos-r-l-rodrigues

Reply:
See #18a70dc